### PR TITLE
mempool, validation: Explain cs_main locking semantics

### PR DESCRIPTION
**Pull request by maflcko**
Both the chain state and the transaction pool are validation specific, but access to them is protected by two locks. The two locks have the following semantics:

* Writing to the chain state or adding transactions to the transaction pool -> Take both `cs_main` and `mempool::cs`
* Reading either or removing transactions from the the transaction pool -> Take only the appropriate lock

### CODE REVIEW

File: src/validation.cpp
Comment (ch4ot1c) [inline src/validation.cpp]: `rading it` -> `reading from it`

File: src/validation.cpp
Comment (ch4ot1c) [inline src/validation.cpp]: `such as` -> `in cases such as`?

**Review by ch4ot1c — CHANGES_REQUESTED** [review]

File: src/txmempool.h
Comment (promag) [inline src/txmempool.h]: `... should be locked before in addition ...`

File: src/txmempool.h
Comment (promag) [inline src/txmempool.h]: `Recursive mutex ...`

File: src/validation.cpp
Comment (promag) [reply src/validation.cpp]: `Recursive mutex to ...`

**Review by promag — COMMENTED** [review]
Concept ACK. Title prefix suggests this is documentation only change, which is not. Some nits though.

File: src/txmempool.h
Comment (MarcoFalke) [reply src/txmempool.h]: Recursive is already clear from the variable name and I plan to make it non-recursive

**Comment by MarcoFalke** [comment]
Fixed up some of the language. Any suggestions for the pull request subject line?

File: src/validation.cpp
Comment (ryanofsky) [inline src/validation.cpp]: In commit "doc: Add comment to cs_main and mempool::cs" (fa64bb42e97b6c6c58e83badca88d386f99dbcd2)

I would say "may also need to be locked" instead of "must also be locked", and add "see CTxMemPool::cs comment for details."

File: src/txmempool.h
Comment (ryanofsky) [inline src/txmempool.h]: In commit "doc: Add comment to cs_main and mempool::cs" (fa64bb42e97b6c6c58e83badca88d386f99dbcd2)

I think making a comparison to read/write locks makes this more confusing than it needs to be. Read locks normally are shared, not exclusive. And the requirement to lock `cs_main` for write access seems too broad, and doesn't give a justification.

I would suggest a comment that just says when you need to lock the mutexes and why:

```c++
/**
 * This mutex needs to be locked when accessing mapTx.
 *
 * @par Relationship to cs_main
 * Locking both mempool.cs and cs_main gives a view of the mempool which is
 * compatible with the current chain tip. To guarantee this, both mutexes need
 * to be locked when adding transactions to the mempool or changing the
 * chain tip. But locking mempool.cs by itself is sufficient for reading
 * transactions from the mempool, or removing transactions from it.
 */
```

**Review by ryanofsky — APPROVED** [review]
utACK fa64bb42e97b6c6c58e83badca88d386f99dbcd2

File: src/validation.cpp
Comment (MarcoFalke) [reply src/validation.cpp]: Done

File: src/bench/mempool_eviction.cpp
Comment (MarcoFalke) [reply src/bench/mempool_eviction.cpp]: Done

File: src/txmempool.h
Comment (MarcoFalke) [reply src/txmempool.h]: It could also be used to guard other members, like prioritization data, no?

Comment (MarcoFalke) [reply src/txmempool.h]: Otherwise done

**Comment by laanwj** [comment]
ACK

File: src/txmempool.h
Comment (ryanofsky) [inline src/txmempool.h]: This comment comes from https://github.com/bitcoin/bitcoin/pull/14963#discussion_r242268194 and I think it is still ok as a description of how the code currently works. But talking to @morcos, @sdaftuar, and you Monday, it sounds like this is not actually how things are intended to work. After that conversation, I think the following might be more accurate:

```c++
/**
 * This mutex needs to be locked when accessing `mapTx` or other members
 * that are guarded by it.
 *
 * @par Consistency guarantees
 *
 * By design, it is guaranteed that:
 *
 * 1. Locking both `cs_main` and `mempool.cs` will give a view of mempool
 *    that is consistent with current chain tip (`chainActive` and
 *    `pcoinsTip`) and is fully populated. Fully populated means that if the
 *    current active chain is missing transactions that were present in a
 *    previously active chain, all the missing transactions will have been
 *    re-added to the mempool and should be present if they meet size and
 *    consistency constraints.
 *
 * 2. Locking `mempool.cs` without `cs_main` will give a view of a mempool
 *    consistent with some chain that was active since `cs_main` was last
 *    locked, and that is fully populated as described above. It is ok for
 *    code that only needs to query or remove transactions from the mempool
 *    to lock just `mempool.cs` without `cs_main`.
 *
 * To provide these guarantees, it is necessary to lock both `cs_main` and
 * `mempool.cs` whenever adding transactions to the mempool and whenever
 * changing the chain tip. It's necessary to keep both mutexes locked until
 * the mempool is consistent with the new chain tip and fully populated.
 *
 * @par Consistency bug
 *
 * The second guarantee above is not currently enforced, but
 * https://github.com/bitcoin/bitcoin/pull/14193 will fix it. No known code
 * in bitcoin currently depends on second guarantee, but it is important to
 * fix for third party code that needs be able to frequently poll the
 * mempool without locking `cs_main` and without encountering missing
 * transactions during reorgs.
 */
```

File: src/validation.cpp
Comment (ryanofsky) [inline src/validation.cpp]: Should say "may also need to be locked" instead of "may also be locked".

**Review by ryanofsky — APPROVED** [review]
utACK fa5c7671aa4c916b191d0423309f89b69568c7f9. The only changes since my previous review were applying suggestions.

File: src/txmempool.h
Comment (MarcoFalke) [reply src/txmempool.h]: Wow, excellent write-up. Will pull that in to replace my clumsy explanation

**Review by ryanofsky — APPROVED** [review]
utACK fa5e373365b8e88fc9894f53f618d3e89f1bec14. Changes since last review: adding longer mempool.cs comment and adding a commit with some new locking annotations.

Also, I don't think this needs to hold up merging, but I think it would be good if a reviewer with more mempool expertise could read the new code comments and ACK if they are correct.

File: src/txmempool.h
Comment (practicalswift) [inline src/txmempool.h]: Changer order of `cs` and `cs_main` here to make locking order consistent?

File: src/txmempool.h
Comment (practicalswift) [inline src/txmempool.h]: Changer order of `cs` and `cs_main` here to make locking order consistent?

**Comment by practicalswift** [comment]
utACK fa5e373365b8e88fc9894f53f618d3e89f1bec14 modulo nits

File: src/validation.cpp
Comment (sdaftuar) [inline src/validation.cpp]: We call the "transaction pool" the "mempool" in so many places in this file, I find the usage of "transaction pool" to be odd...

**Review by sdaftuar — APPROVED** [review]
utACK fa5e373365b8e88fc9894f53f618d3e89f1bec14

**Pull request merged**
